[SPARK-26729][K8S] Make image names under test configurable

rvesse opened the pull request:
## What changes were proposed in this pull request?

Allow specifying system properties to customise the image names for the images used in the integration testing.  Useful if your CI/CD pipeline or policy requires using a different naming format.

This is one part of addressing SPARK-26729, I plan to have a follow up patch that will also make the names configurable when using `docker-image-tool.sh`

## How was this patch tested?

Ran integration tests against custom images generated by our CI/CD pipeline that do not follow Spark's existing hardcoded naming conventions using the new system properties to override the image names appropriately:

```
mvn clean integration-test -pl :spark-kubernetes-integration-tests_${SCALA_VERSION} \
            -Pkubernetes -Pkubernetes-integration-tests \
            -P${SPARK_HADOOP_PROFILE} -Dhadoop.version=${HADOOP_VERSION} \
            -Dspark.kubernetes.test.sparkTgz=${TARBALL} \
            -Dspark.kubernetes.test.imageTag=${TAG} \
            -Dspark.kubernetes.test.imageRepo=${REPO} \
            -Dspark.kubernetes.test.namespace=${K8S_NAMESPACE} \
            -Dspark.kubernetes.test.kubeConfigContext=${K8S_CONTEXT} \
            -Dspark.kubernetes.test.deployMode=${K8S_TEST_DEPLOY_MODE} \
            -Dspark.kubernetes.test.jvmImage=apache-spark \
            -Dspark.kubernetes.test.pythonImage=apache-spark-py \
            -Dspark.kubernetes.test.rImage=apache-spark-r \
            -Dtest.include.tags=k8s
...
[INFO] --- scalatest-maven-plugin:1.0:test (integration-test) @ spark-kubernetes-integration-tests_2.12 ---
Discovery starting.
Discovery completed in 230 milliseconds.
Run starting. Expected test count is: 15
KubernetesSuite:
- Run SparkPi with no resources
- Run SparkPi with a very long application name.
- Use SparkLauncher.NO_RESOURCE
- Run SparkPi with a master URL without a scheme.
- Run SparkPi with an argument.
- Run SparkPi with custom labels, annotations, and environment variables.
- Run extraJVMOptions check on driver
- Run SparkRemoteFileTest using a remote data file
- Run SparkPi with env and mount secrets.
- Run PySpark on simple pi.py example
- Run PySpark with Python2 to test a pyfiles example
- Run PySpark with Python3 to test a pyfiles example
- Run PySpark with memory customization
- Run in client mode.
- Start pod creation from template
Run completed in 8 minutes, 33 seconds.
Total number of tests run: 15
Suites: completed 2, aborted 0
Tests: succeeded 15, failed 0, canceled 0, ignored 0, pending 0
All tests passed.
```

rvesse committed:
[SPARK-26729][K8S] Make image names under test configurable

Allow specifying system properties to customise the image names for the
images used in the integration testing.  Useful if your CI/CD pipeline
or policy requires using a different naming format.

rvesse committed:
[SPARK-26729][K8S] Propagate new config to int test

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #102560 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102560/testReport)** for PR 23846 at commit [`d9871a6`](https://github.com/apache/spark/commit/d9871a6e1956fbc96f327f5f423e453ae30f1170).

SparkQA commented:
**[Test build #102560 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102560/testReport)** for PR 23846 at commit [`d9871a6`](https://github.com/apache/spark/commit/d9871a6e1956fbc96f327f5f423e453ae30f1170).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102560/
Test FAILed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/8120/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/8120/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8120/
Test PASSed.

rvesse committed:
[SPARK-26729][K8S] Fix scala style violations

SparkQA commented:
**[Test build #102562 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102562/testReport)** for PR 23846 at commit [`d45b136`](https://github.com/apache/spark/commit/d45b136a0a5444a2e1f8d2beeb376c1bd4c5e56c).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/8122/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/8122/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8122/
Test PASSed.

SparkQA commented:
**[Test build #102562 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102562/testReport)** for PR 23846 at commit [`d45b136`](https://github.com/apache/spark/commit/d45b136a0a5444a2e1f8d2beeb376c1bd4c5e56c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102562/
Test FAILed.

felixcheung commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/SecretsTestsSuite.scala:
```diff
@@ -72,7 +72,7 @@ private[spark] trait SecretsTestsSuite { k8sSuite: KubernetesSuite =>
           checkSecrets(executorPod)
         },
-        appArgs = Array("1000") // give it enough time for all execs to be visible
+        appArgs = Array("10000") // give it enough time for all execs to be visible
```
why change these?

rvesse replied:
On a slow cluster these timeouts are too short.  When testing this against our custom named images on a low spec dev cluster this test repeatedly failed because the executors simply didn't launch fast enough

skonto replied:
Yes that is correct on our ci we have increased that too, but on minikube is fine.

rvesse replied:
Removed this change from this PR, will submit separately

skonto commented on resource-managers/kubernetes/integration-tests/pom.xml:
```diff
@@ -148,6 +148,9 @@
             <spark.kubernetes.test.master>${spark.kubernetes.test.master}</spark.kubernetes.test.master>
             <spark.kubernetes.test.namespace>${spark.kubernetes.test.namespace}</spark.kubernetes.test.namespace>
             <spark.kubernetes.test.serviceAccountName>${spark.kubernetes.test.serviceAccountName}</spark.kubernetes.test.serviceAccountName>
+            <spark.kubernetes.test.jvmImage>${spark.kubernetes.test.jvmImage}</spark.kubernetes.test.jvmImage>
```
So far you can configure the image with `--image-repo` and `--image-tag` but the name part was hardcoded.

srowen replied:
I know this isn't new, but do all of these have to be system properties vs some command line argument somewhere? Probably, and it's fine, just checking

rvesse replied:
I think we can add corresponding arguments to the `integration-test.sh` script similar to what is supported currently per @skonto's comment.  I will update with that and other requested changes next week

rvesse replied:
@srowen Now also exposed as CLI arguments to the integration test script and documented accordingly in the README

SparkQA commented:
**[Test build #103054 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103054/testReport)** for PR 23846 at commit [`20bc1c0`](https://github.com/apache/spark/commit/20bc1c0062f5bef87e1cb580b8dfcba3f735b96c).

rvesse committed:
[SPARK-26729][K8S] Add CLI options for image name customisation

Adds corresponding CLI options to dev/dev-run-integration-tests.sh to
expose the ability to customise the image names when running via the
script.

Update integration test README appropriately

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/8516/

SparkQA commented:
**[Test build #103055 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103055/testReport)** for PR 23846 at commit [`00fcf90`](https://github.com/apache/spark/commit/00fcf90c1cfef965df20ea576dea2231e14d7558).

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/8516/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8516/
Test PASSed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/8517/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/8517/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8517/
Test PASSed.

SparkQA commented:
**[Test build #103054 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103054/testReport)** for PR 23846 at commit [`20bc1c0`](https://github.com/apache/spark/commit/20bc1c0062f5bef87e1cb580b8dfcba3f735b96c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103054/
Test PASSed.

SparkQA commented:
**[Test build #103055 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103055/testReport)** for PR 23846 at commit [`00fcf90`](https://github.com/apache/spark/commit/00fcf90c1cfef965df20ea576dea2231e14d7558).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103055/
Test PASSed.

vanzin commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:
```diff
@@ -108,9 +108,12 @@ class KubernetesSuite extends SparkFunSuite
-    image = testImageRef("spark")
-    pyImage = testImageRef("spark-py")
-    rImage = testImageRef("spark-r")
+    image = testImageRef(Option(System.getProperty(CONFIG_KEY_IMAGE_JVM))
```
`sys.props.getOrElse`

rvesse replied:
Updated to use simpler Scala convenience method

vanzin reviewed (commented):
Small suggestion otherwise looks ok.

holdenk commented:
LGTM pending vanzin's suggestion

rvesse committed:
[SPARK-26729][K8S] Use scala convinience method

rvesse commented:
Updated per reviewer comments

SparkQA commented:
**[Test build #103731 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103731/testReport)** for PR 23846 at commit [`b7d2bfb`](https://github.com/apache/spark/commit/b7d2bfbf61fe562a2e9694428b8b18873a3c604b).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/9113/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/testing-k8s-prb-make-spark-distribution-unified/9113/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9113/
Test PASSed.

SparkQA commented:
**[Test build #103731 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103731/testReport)** for PR 23846 at commit [`b7d2bfb`](https://github.com/apache/spark/commit/b7d2bfbf61fe562a2e9694428b8b18873a3c604b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

vanzin commented:
Merging to master.

liu-sheng commented on resource-managers/kubernetes/integration-tests/dev/dev-run-integration-tests.sh:
```diff
@@ -119,4 +139,15 @@ then
+PYTHON_IMAGE_NAME=${PYTHON_IMAGE_NAME:-${BASE_IMAGE_NAME}-py}
+R_IMAGE_NAME=${R_IMAGE_NAME:-${R_IMAGE_NAME}-r}
+
+properties+=(
```
it seems this `roperties+=(...` can not work, I have added a echo statement before the `mvn` command, it shows the `properties` didn't contain the image properties, see following log message:

```
+ properties=(${properties[@]} -Dtest.exclude.tags=$EXCLUDE_TAGS)
+ BASE_IMAGE_NAME=spark
+ JVM_IMAGE_NAME=spark
+ PYTHON_IMAGE_NAME=spark-py
+ R_IMAGE_NAME=-r
+ properties+=(-Dspark.kubernetes.test.jvmImage=$JVM_IMAGE_NAME -Dspark.kubernetes.test.pythonImage=$PYTHON_IMAGE_NAME -Dspark.kubernetes.test.rImage=$R_IMAGE_NAME)
+ echo -Dspark.kubernetes.test.sparkTgz=/home/zuul/src/github.com/liu-sheng/spark/spark-3.0.0-SNAPSHOT-bin-2.7.4.tgz
```

rvesse replied:
@liu-sheng What did you echo?  An `echo $properties` will only echo the first item in the array, an `echo ${properties[@]}` echos the entire array.  It looks like you only echo'd the first item in the array

Note that you can invoke the script with `sh -x dev/dev-run-integration-tests.sh` to get detailed Bash debug output.

liu-sheng replied:
> @liu-sheng What did you echo? An `echo $properties` will only echo the first item in the array, an `echo ${properties[@]}` echos the entire array. It looks like you only echo'd the first item in the array
> 
> Note that you can invoke the script with `sh -x dev/dev-run-integration-tests.sh` to get detailed Bash debug output.

Oh, my fault, let me try, thank you for teaching me :)

liu-sheng replied:
BTW, the L145 `R_IMAGE_NAME=${R_IMAGE_NAME:-${R_IMAGE_NAME}-r}`, is the default `R_IMAGE_NAME` var name should be `BASE_IMAGE_NAME` ? if I do not specify the R image with `--r-image-name`, the R image name will be `-r` as default.

rvesse replied:
Oops, then that's a bug that made it past the review.  Opened PR #24449

rvesse replied:
@liu-sheng This should now be fixed in latest master

liu-sheng replied:
@rvesse thank you :)
